services/packages/arch/db_manager.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package arch
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"strings"
+
+	"code.gitea.io/gitea/models/db"
+	pkg_model "code.gitea.io/gitea/models/packages"
+	"code.gitea.io/gitea/models/repo"
+	"code.gitea.io/gitea/models/user"
+	"code.gitea.io/gitea/modules/context"
+	"code.gitea.io/gitea/modules/json"
+	"code.gitea.io/gitea/modules/packages"
+	"code.gitea.io/gitea/modules/packages/arch"
+	"code.gitea.io/gitea/modules/storage"
+	pkg_service "code.gitea.io/gitea/services/packages"
+)
+
+type UpdateMetadataParameters struct {
+	User *user.User
+	Md   *arch.Metadata
+}
+
+// This function gets existing package metadata for provided version present,
+// combines architecture and distribution info and creates new metadata with
+// combined set of parameters.
+func UpdateMetadata(ctx *context.Context, p *UpdateMetadataParameters) error {
+	ver, err := pkg_model.GetVersionByNameAndVersion(ctx, p.User.ID, pkg_model.TypeArch, p.Md.Name, p.Md.Version)
+	if err != nil {
+		return err
+	}
+
+	var currmd arch.Metadata
+	err = json.Unmarshal([]byte(ver.MetadataJSON), &currmd)
+	if err != nil {
+		return err
+	}
+
+	currmd.Arch = arch.UnifiedList(currmd.Arch, p.Md.Arch)
+	currmd.DistroArch = arch.UnifiedList(currmd.DistroArch, p.Md.DistroArch)
+
+	b, err := json.Marshal(&currmd)
+	if err != nil {
+		return err
+	}
+
+	ver.MetadataJSON = string(b)
+
+	return pkg_model.UpdateVersion(ctx, ver)
+}
+
+// Parameters required to save new arch package.
+type SaveFileParams struct {
+	Creator  *user.User
+	Owner    *user.User
+	Metadata *arch.Metadata
+	Data     []byte
+	Filename string
+	Distro   string
+	IsLead   bool
+}
+
+// This function creates new package, version and package_file properties in
+// database, and writes blob to file storage. If package/version/blob exists it
+// will overwrite existing data. Package id and error will be returned.
+func SaveFile(ctx *context.Context, p *SaveFileParams) (int64, error) {
```

Comment:
I see no benefit of this helper method.

Reply:
@KN4CK3R 
I save 2 files (3 if signature provided) for each package. Which technique should i use to reduce code repetition?